vip-helpers/class-vip-media-library-caching.php

```diff
@@ -0,0 +1,210 @@
+<?php
+/**
+ * Media library caching helpers.
+ * 
+ * @package vip-helpers
+ */
+
+ // phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery
+
+namespace Automattic\VIP\Helpers;
+
+/**
+ * VIP Media Library Caching class.
+ */
+class VIP_Media_Library_Caching {
+	public const MINIMUM_WORDPRESS_VERSION          = '6.4';
+	public const AVAILABLE_MIME_TYPES_CACHE_KEY     = 'vip_available_mime_types_';
+	public const USING_DEFAULT_MIME_TYPES_CACHE_KEY = 'vip_using_default_mime_types_';
+	public const MAX_POSTS_TO_QUERY                 = 100000;
+
+	/**
+	 * Class constructor.
+	 */
+	public function __construct() {
+		global $wp_version;
+
+		if ( ! defined( 'VIP_DISABLE_MIME_TYPE_CACHING' ) &&
+			isset( $wp_version ) &&
+			version_compare( $wp_version, self::MINIMUM_WORDPRESS_VERSION, '>=' ) ) {
+			$this->enable_post_mime_types_caching();
+		}
+	}
+
+	/**
+	 * Enable MIME type caching.
+	 */
+	private function enable_post_mime_types_caching() {
+		add_filter( 'pre_get_available_post_mime_types', array( $this, 'get_cached_post_mime_types' ), 10, 2 );
+		add_action( 'add_attachment', array( $this, 'update_post_mime_types_cache_on_add', 10, 1 ) );
+		add_action( 'attachment_updated', array( $this, 'update_post_mime_types_cache_on_edit' ), 10, 3 );
+		add_action( 'delete_attachment', array( $this, 'update_post_mime_types_cache_on_delete' ), 10, 3 );
+	}
+
+	/**
+	 * Get cached results for get_available_post_mime_types() to avoid a query on every page load.
+	 * 
+	 * @param string[]|null $filtered_mime_types An array of MIME types. Default null.
+	 * @param string        $type                The post type name. Usually 'attachment' but can be any post type.
+	 * @return array An array of MIME types.
+	 */
+	public function get_cached_post_mime_types( $filtered_mime_types, $type ) {
```

Comment:
Do we care about running this filter if `$type` isn't `attachment`? My understanding is that the expensive query stems from https://github.com/WordPress/WordPress/blob/b0da2934b925077ab95cd36cad4217ed214665b1/wp-admin/includes/post.php#L1390

Reply:
I added `$type` to preserve the method signature and filter definition from:
https://github.com/WordPress/WordPress/blob/b0da2934b925077ab95cd36cad4217ed214665b1/wp-includes/post.php#L7937

What you pointed out is the only place where that function is called in core, and 'attachment' is hard coded. But I still preferred to make the distinction to make the code more resilient if, for some (unlikely) reason, core decided to call that function for a different post type at some point.

Reply:
Sorry for any confusion, to clarify -- I meant can we bail early if `$type` isn't attachment?

Reply:
Ok, I get that. It's only that the current structure would allow caching of MIME types for attachments OR any other post types that might come in the future through that filter. If we restrict it to only attachments, a lot of the logic involving variable cache buckets according to the `$type` variable could also be removed from the other places, and `attachment` could be hard coded. But I leave the final call up to you. 🙂

Reply:
Given that the hooks we are using are attachment specific, it'd make sense to only cache for attachments:
```
add_action( 'add_attachment', array( __CLASS__, 'update_post_mime_types_cache_on_add', 10, 1 ) );
		add_action( 'attachment_updated', array( __CLASS__, 'update_post_mime_types_cache_on_edit' ), 10, 3 );
		add_action( 'delete_attachment', array( __CLASS__, 'update_post_mime_types_cache_on_delete' ), 10, 2 );
		```

Reply:
I removed all the post-type-based logic, and it will only cache attachments now.